Add role-based mutes and multi-server support

Dogel-ai opened the pull request.

Nuhuhb committed:
feat: apply configured role(s) on user timeout

Dogel-ai committed:
feat: apply configured role(s) on user timeout

Dogel-ai committed:
Merge branch 'feature/non-native-roll-support' of https://github.com/Dogel-ai/Amazake into feature/non-native-roll-support

Dogel-ai committed:
feat: store user timeouts in redis

Dogel-ai committed:
feat: add periodical unmute and multi-sever non-native-roll-support

Kyrielight commented on config.py:
```diff
@@ -15,8 +15,13 @@
         Validator("bot_token", must_exist=True, is_type_of=str),
+        # Redis settings
+        Validator("redis_host", must_exist=True, is_type_of=str),
+        Validator("redis_port", must_exist=True, is_type_of=int),
```
Good start! As part of the MVP, we should also have support for a Redis username and password (and lack thereof). 

(For context: The production network actually does have an authenticated Redis instance for tracking leaderboards)

Kyrielight replied:
Good start! As part of the MVP, we should also have support for a Redis username and password (and lack thereof).

(For context: The production network actually does have an authenticated Redis instance for tracking leaderboards)

Kyrielight commented on config.py:
```diff
@@ -15,8 +15,13 @@
+        Validator("redis_host", must_exist=True, is_type_of=str),
+        Validator("redis_port", must_exist=True, is_type_of=int),
         # Roulette settings
+        Validator("unmute_rate",  is_type_of=int),
```
nit: Let's be pedantic and change this to match the functionality: `roulette_unmute_rate`. The unmute currently only applies to everything under the /roulette extension (while this bot at a greater level is technically generic)

(You'll need to change the underlying config methods too accordingly)

Kyrielight commented on config.py:
```diff
@@ -46,6 +51,27 @@ def bot_token() -> str:
+def unmute_rate() -> int:
+    """
+    :return: Time in minutes between each unmute loop, as an integer.
+    """
```
You can omit the documentation here and move it into `//extensions/roulette/config/config.py` (since it will be a Roulette-only applicable configuration)

Kyrielight commented on config/settings.toml:
```diff
@@ -25,7 +40,7 @@ roulette_moderator_roles = ["<list_of_role_ids>"]
 # Reserved superuser role for future use.
 # Administrators are implicitly moderators.
-roulette_administrator_users = ["<list_of_role_ids>"]
+roulette_administrator_users = ["<list_of_user_ids>"]
```
aha thanks

Kyrielight commented on extensions/roulette/extension.py:
```diff
@@ -15,3 +16,7 @@ async def setup(bot: Bot) -> None:
     await bot.add_cog(Roll(bot))
     logger.info("Loaded Roll extension")
+
+    logger.info("Loading Unmute extension")
+    await bot.add_cog(Unmute(bot))
+    logger.info("Loaded Unmute extension")
```
Note: Let's actually load this *before* the roll - if the unmute fails to load, we shouldn't load the rolls either (for obvious reasons)

Kyrielight commented on extensions/roulette/config/redis_config.py:
```diff
@@ -0,0 +1,9 @@
+    port = config.redis_port()
+)
+
+redis_client = redis.Redis(connection_pool=pool)
```
Slight overkill but:

The Redis connection is defined at the bot-scope, not the extension scope (good!). But this means access to Redis should also be bot-scope, not only within the confines of the `roulette` extension.

So what I would recommend is:

1. Make a root-level folder called "Redis" and move this file there (don't name it "redis.py" for collissions)
2. In that file, internally (using `_`) create a `ConnectionPool` object (so as to reuse Redis connections)
3. Expose a method in that file called `get_redis()` (or something similar) that returns a Redis client that uses the pool:
```py
def get_redis():
  return redis.Redis(connection_pool=_pool)
```
This DI pattern allows us to reuse the connection pool instead of spamming new threads each time, but ensures client uniqueness. Do NOT expose the connection pool object itself.

4. Under `extensions/roulette/db` (or a similar name referring to a database), create an API to abstract away talking to Redis. This API internally will use the global pool:
```py
# In extensions/roulette/db/redis_db.py, for example:

_redis_client = get_redis()

# Change the parameters here to whatever you need
def add_timeout():
  pass # implement

def remove_timeout():
  pass # implement
```

5. Use this API inside `roulette/timeout` and `roulette/unmute`

Kyrielight commented on extensions/roulette/roll/cog.py:
```diff
@@ -3,9 +3,10 @@
 
 from asyncio import sleep
-from datetime import timedelta
+from datetime import timedelta, datetime
```
nit: Please maintain alphabetical order here

Kyrielight commented on extensions/roulette/roll/action.py:
```diff
@@ -16,6 +16,9 @@
 
+class Role:
+    def __init__(self, id: int):
+        self.id = id
```
This doesn't appear to be used anywhere?

Kyrielight commented on extensions/roulette/unmute/action.py:
```diff
@@ -0,0 +1,7 @@
+
+class Role:
+    def __init__(self, id: int):
+        self.id = id
```
ditto: This doesn't appear to be used? (And like below, prefer to use discord.py's id role directly)

Kyrielight commented on extensions/roulette/roll/cog.py:
```diff
@@ -99,6 +100,15 @@ def _is_admin(self, user: User | Member) -> bool:
+            role = target.guild.get_role(int(timeout_id))
+
+            if role:
+                await target.add_roles(role, reason=f"Timed out for {duration_label} via Roulette")
```
nit: Let's be semantic and indicate it's more like "adding role for duration timeout" -the timeout itself may occur in multiple (overlapping) fashions.

Kyrielight commented on extensions/roulette/roll/cog.py:
```diff
@@ -162,6 +172,20 @@ async def _determine_targets(self, message: Message) -> Set[Member]:
+                              message: Message,
+                              target: Member) -> int:
+
+        key = "roulette_timeout_live_" + str(message.guild.id)
```
note: This key should be a file-level constant (which will be defined in the roulette redis file following the comment above).

(Keep this `_upload_timeout` method; just move the `zadd` method into the separate API)

Kyrielight commented on extensions/roulette/unmute/cog.py:
```diff
@@ -0,0 +1,88 @@
+    async def cog_command_error(self, ctx, error: Exception) -> None:
+        self.logger.warning(str(error))
+
+    @tasks.loop(minutes=1)
```
Just curious: can this not be set directly from config? (I would guess so since it's a decorator function).

If so: If possible, set this to `sys.MAXSIZE` instead - Redis requests (depending on deployment situation) can be charged per command, so we don't to send any more requests than necessary until the config is loaded.

Dogel-ai replied:
If you mean the @tasks.loop(), it's the amount of time it waits between loops.

Kyrielight replied:
Yeah, the value should preferably be set to `sys.MAXSIZE` here first. Since Redis requests can cost money, we want to minimise the number of calls until the config is loaded (or if there's an unexpected error in loading)

Kyrielight commented on extensions/roulette/unmute/cog.py:
```diff
@@ -0,0 +1,88 @@
+        self.logger.info("Loaded Unmute cog")
+
+    async def cog_command_error(self, ctx, error: Exception) -> None:
+        self.logger.warning(str(error))
```
nit: use `ERROR` logging level here

Kyrielight commented on extensions/roulette/unmute/cog.py:
```diff
@@ -0,0 +1,88 @@
+                break
+
+        for key in guild_keys:
+            guild_ids.add(int(key[22:]))
```
nit: Don't use [22:] here; it's not particularly readable. Prefer to reference the pattern again if you can.

Kyrielight reviewed (changes requested):
Great start! Thanks for contributing.

I've left an initial round of comments. Some of these will change the structure relatively seriously, so I won't go too deep into all files until major changes are made.

Dogel-ai committed:
feat: add api to abstract redis

* Add username and password for redis authentication.

* Load unmute before roll extension

* Remove abandoned files

* Add redis key mod to config

* Chores

Kyrielight commented on Redis/redis_config.py:
```diff
@@ -0,0 +1,16 @@
+import config
```
nit: `redis` (not `Redis`) in the file path

Dogel-ai replied:
Changing it to redis would conflict with the package redis. I've renamed it redis_interface to keep them separate.

Kyrielight commented on config.py:
```diff
@@ -15,8 +15,16 @@
         # Roulette settings
+        Validator("roulette_unmute_rate",  is_type_of=int),
         Validator("roulette_channels", must_exist=True, is_type_of=list, len_min=1),
+        Validator("roulette_timeout_roles", must_exist=True, is_type_of=list, len_min=1),
```
note: This will be a breaking change, and it's probably better to ensure the bot still supports a no-role version (pure timeout only). Let's remove `must_exist=True` (and `len_min`) here - the actual role application code can just iterate over an empty list and no-op.

Dogel-ai replied:
While this works fine, it disables some logging messages after the timeout passed. Shouldn't be a problem with large data sets, though I will mark it as a todo.

Kyrielight commented on config.py:
```diff
@@ -46,6 +54,42 @@ def bot_token() -> str:
+    return _settings.get("redis_port")
+
+
+def redis_username() -> str:
+    """
+    :return: Username used for redis connection, as a string.
+    """
+    return _settings.get("redis_username") or None
+
+
+def redis_password() -> str:
+    """
+    :return: Password used for redis connection, as a string.
+    """
+    return _settings.get("redis_password") or None
+
+
+def redis_key_const() -> str:
+    return _settings.get("redis_key_const")
+
+
+def roulette_unmute_rate() -> int:
+    return _settings.get("roulette_unmute_rate")
+
+
```
The return type here for these fields is `Optional[str]` or `Optional[int]` respectively (they can return `None`)

Note to self: I should add a static analysis checker to this project eventually. Or just change it to TypeScript...

Kyrielight commented on config.py:
```diff
@@ -54,6 +98,10 @@ def roulette_channels() -> List[str]:
     return _settings.get("roulette_channels")
 
 
+def roulette_timeout_roles() -> List[str]:
```
(note: This will become `Optional[List[str]]` after changes)

Kyrielight commented on extensions/roulette/database/action.py:
```diff
@@ -0,0 +1,23 @@
+    role_ids_conf = config.timeout_roles()
+    role_ids = set()
+
+    for role_id_conf in role_ids_conf:
```
`for role_id in role_ids_conf` (the iteration is over role IDs, not configurations). Maybe also make the set `timeout_roles` instead of `role_ids` for clarity

Kyrielight commented on extensions/roulette/database/action.py:
```diff
@@ -0,0 +1,23 @@
+
+    for role_id_conf in role_ids_conf:
+        role = guild.get_role(int(role_id_conf))
+        if role:
```
note: Could you add a TODO here that says "check if a call to `fetch_roles` is needed? I'm concerned there may be a scenario where the `guild` object doesn't actually contain the roles, so this doesn't do anything

Kyrielight commented on extensions/roulette/database/action.py:
```diff
@@ -0,0 +1,23 @@
+from typing import Set
+
+
+class Role:
```
Question: Is it possible to use discord.py's inbuilt `Role` class instead? There doesn't seem to be a significant reason to redefine one ourselves. Then `get_timeout_roles` can just become a public method (I'd also avoid using `fetch` since Discord.py uses that to imply API calls)

Dogel-ai replied:
While it's not really possible to use the `Role` class, I believe that using the `Object` class and passing the id would be the solution. (It works so far)

`return {Object(id=timeout_role) for timeout_role in timeout_roles}`

Dogel-ai replied:
Scratch that, it broke as soon as I posted the comment. I'll try to figure out why and possibly fix it

Kyrielight commented on extensions/roulette/database/redis_db.py:
```diff
@@ -0,0 +1,95 @@
+            await target.add_roles(role, reason=f"Adding role for {duration_label} timeout")
+
+
+async def add_timeout(target, duration_label, duration, message):
+    await _upload_timeout(duration, message, target)
```
imo there's no particular need to have a wrapper here; can just make `_upload_timeout` called `add_timeout` directly.

Kyrielight commented on extensions/roulette/database/redis_db.py:
```diff
@@ -0,0 +1,95 @@
+# TODO: Check for successful update
+async def _upload_timeout(duration: timedelta,
+                          message: Message,
+                          target: Member) -> int:
```
nitpicky, but could we add an Enum for return values, like SUCCESS and FAILURE? imo there's no need to expose the internal workings of Redis to client calling code (and make them have to understand what Redis respond codes are)

Kyrielight commented on extensions/roulette/database/redis_db.py:
```diff
@@ -0,0 +1,95 @@
+            break
+
+    for key in guild_keys:
+        guild_ids.add(int(key[22:]))
```
Don't use [22:] here; it's not semantic. (Also, I think this will break because the prefix is configurable?)

Kyrielight commented on extensions/roulette/database/redis_db.py:
```diff
@@ -0,0 +1,95 @@
+    return guild_ids
+
+
+async def _fetch_candidates(guild_id) -> Set[int]:
```
nit: `candidates` is a bit vague here. These specifically are muted users who may be unmutable.

Kyrielight commented on extensions/roulette/roll/cog.py:
```diff
@@ -185,14 +195,14 @@ async def _timeout(self,
-            await message.reply("Sorry, something went wrong. Please roll again!")
-            return
+        if not duration > timedelta(days=28):
+            await target.timeout(duration, reason=f"Timed out for {duration_label} via Roulette")
```
Let's be clear this is a Discord native timeout

Dogel-ai committed:
feat: add support for no-role version

- change redis parameters to optional

- naming changes

- remove unnecessary wrappers

- add enum for _upload_timeout return values

- fix _get_guild_ids to work with configurable prefix

- change unmute loop interval to minimise calls to redis

Dogel-ai committed:
Merge pull request #1 from Dogel-ai/feature/non-native-roll-support

feat: add support for no-role version

Kyrielight commented on extensions/roulette/roll/cog.py:
```diff
@@ -205,4 +217,4 @@ async def _timeout(self,
                                              duration_label=duration_label))
 
-        stats.timeout_record_stats(duration, message)
+        #stats.timeout_record_stats(duration, message)
```
Last question, sorry: Is this intentionally disabled? Doesn't seem like there's any particular reason why `stats` is disabled.

Dogel-ai replied:
I've disabled it while testing because it threw an error each time and just forgot to re-enable it. Just pushed a fix

Kyrielight reviewed (approved):
Overall LGTM - there are some nitpicks, but they're not too important and can be fixed in another PR.

Dogel-ai committed:
fix: re-enable stats
